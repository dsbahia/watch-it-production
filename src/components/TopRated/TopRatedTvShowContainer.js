import React, { useEffect, useState } from "react";
import { toast } from "react-hot-toast";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCircleChevronDown } from "@fortawesome/free-solid-svg-icons";
import TopRatedTvShow from "./TopRatedTvShow";
import api from "../../api/api";
import "../../styles/topratedmoviescontainer.css";
import WatchItLogo from "../WatchItLogo";

function TopRatedTvShowContainer() {
  const [moviesData, setMoviesData] = useState([]);
  const [movieTrailers, setMovieTrailers] = useState({});
  const [maxResults, setMaxResults] = useState(9);

  useEffect(() => {
    async function fetchTopRatedTvShow() {
      try {
        const data = await api.topRatedTvShows();
        setMoviesData(data.results);

        const trailers = {};
        await Promise.all(
          data.results.map(async (tv) => {
            const trailerData = await api.tvTrailer(tv.id);

            if (trailerData.results && Array.isArray(trailerData.results)) {
              const trailer = trailerData.results.find(
                (item) => item.type === "Trailer",
              );
              if (trailer) {
                const trailerUrl = `https://www.youtube.com/watch?v=${trailer.key}`;
                trailers[tv.id] = trailerUrl;
              }
            } else {
              console.error("Unexpected trailerData structure:", trailerData);
            }
          }),
        );
        setMovieTrailers(trailers);
      } catch (error) {
        const errorMsg = "An error occurred. Please try again later.";
        toast.error(errorMsg, {
          duration: 5000,
        });
      }
    }
    fetchTopRatedTvShow();
  }, []);

  const handleShowMore = () => {
    setMaxResults(maxResults + 6);
  };
  return (
    <div>
      <WatchItLogo />
      <div data-testid="tv-list" className="top-rated-container">
        <div className="top-rated-movie-title">Top Rated Tv Shows</div>
        <div className="top-rated-results">
          {moviesData.slice(0, maxResults).map((data) => (
            <div key={data.id} className="top-rated-item">
              <TopRatedTvShow
                title={data.name}
                posterpath={data.poster_path}
                tvId={data.id}
                tvshowTrailer={movieTrailers[data.id]}
              />
            </div>
          ))}
        </div>{" "}
        {maxResults < moviesData.length && (
          <button
            type="button"
            className="show-more-button"
            onClick={handleShowMore}
          >
            <FontAwesomeIcon
              icon={faCircleChevronDown}
              size="2xl"
              style={{ color: "#393E46" }}
            />
          </button>
        )}
      </div>
    </div>
  );
}

export default TopRatedTvShowContainer;
