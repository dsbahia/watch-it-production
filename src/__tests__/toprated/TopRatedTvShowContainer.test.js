import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import TopRatedTvShowContainer from "../../components/TopRated/TopRatedTvShowContainer";

describe("TopRatedTvShowContainer", () => {
  it("should render without errors", () => {
    render(
      <MemoryRouter>
        <TopRatedTvShowContainer />
      </MemoryRouter>,
    );
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <TopRatedTvShowContainer />
      </MemoryRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("displays a list of top rated movies", async () => {
    render(
      <MemoryRouter>
        <TopRatedTvShowContainer />
      </MemoryRouter>,
    );

    const movieList = await screen.findByTestId("tv-list");

    expect(movieList).toBeInTheDocument();
  });
});
