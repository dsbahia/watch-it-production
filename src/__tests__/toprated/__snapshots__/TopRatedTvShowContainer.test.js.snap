// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopRatedTvShowContainer matches the snapshot 1`] = `
<DocumentFragment>
  <div>
    <a
      href="/"
    >
      <img
        alt="Watch It Logo"
        class="watch-it-logo"
        src="watch-it.png"
      />
    </a>
    <div
      class="top-rated-container"
      data-testid="tv-list"
    >
      <div
        class="top-rated-movie-title"
      >
        Top Rated Tv Shows
      </div>
      <div
        class="top-rated-results"
      />
       
    </div>
  </div>
</DocumentFragment>
`;
